src/workerd/util/state-machine.h

```diff
@@ -0,0 +1,3897 @@
+  KJ_WARN_UNUSED_RESULT kj::Maybe<To&> transitionFromToWith(Producer&& producer) KJ_LIFETIMEBOUND
+    requires(_::isOneOf<From, States...>) && (_::isOneOf<To, States...>)
+  {
+    requireUnlocked();
+
+    if (!is<From>()) {
+      return kj::none;
+    }
+
+    // Lock while evaluating producer to ensure atomicity
+    kj::Maybe<To> maybeNewState;
+    {
+      auto lock = acquireTransitionLock();
+      maybeNewState = producer(state.template get<From>());
+    }
+
+    KJ_IF_SOME(newState, maybeNewState) {
+      return state.template init<To>(kj::mv(newState));
+    }
+    return kj::none;
+  }
+
+  // ---------------------------------------------------------------------------
+  // State Introspection
+  // ---------------------------------------------------------------------------
+
+  // Get the name of the current state (requires states to have NAME member)
+  kj::StringPtr currentStateName() const {
+    if (!isInitialized()) {
+      return "(uninitialized)"_kj;
+    }
+    kj::StringPtr result = "(unknown)"_kj;
+    visit([&result](const auto& s) {
+      using S = kj::Decay<decltype(s)>;
+      result = _::getStateName<S>();
+    });
+    return result;
+  }
+
+  // ---------------------------------------------------------------------------
+  // Visitor Pattern
+  // ---------------------------------------------------------------------------
+
+  // Visit the current state with a generic lambda.
+  // The lambda must be able to accept any state type.
+  //
+  // Usage:
+  //   state.visit([](auto& s) {
+  //     // s is a reference to the current state
+  //   });
+  //
+  // Or with explicit type handling:
+  //   state.visit([](auto& s) {
+  //     using S = kj::Decay<decltype(s)>;
+  //     if constexpr (kj::isSameType<S, Readable>()) { ... }
+  //   });
+  template <typename Visitor>
+  decltype(auto) visit(Visitor&& visitor) {
+    return visitImpl(kj::fwd<Visitor>(visitor), std::index_sequence_for<States...>{});
+  }
+
+  template <typename Visitor>
+  decltype(auto) visit(Visitor&& visitor) const {
+    return visitConstImpl(kj::fwd<Visitor>(visitor), std::index_sequence_for<States...>{});
+  }
+
+  // ---------------------------------------------------------------------------
+  // Interop
+  // ---------------------------------------------------------------------------
+
+  // Access the underlying kj::OneOf for interop with existing code
+  StateUnion& underlying() KJ_LIFETIMEBOUND {
+    return state;
+  }
+  const StateUnion& underlying() const KJ_LIFETIMEBOUND {
+    return state;
+  }
+
+  // For use with KJ_SWITCH_ONEOF
+  auto _switchSubject() & {
+    requireInitialized();
+    return state._switchSubject();
+  }
+  auto _switchSubject() const& {
+    requireInitialized();
+    return state._switchSubject();
+  }
+  auto _switchSubject() && {
+    requireInitialized();
+    return kj::mv(state)._switchSubject();
+  }
+
+ protected:
+  StateUnion state;
+  mutable uint32_t transitionLockCount = 0;
```

Comment:
In our codebase, `const` functions mean they are thread-safe. It doesn't look like the `const` functions which mutate this member are actually thread-safe, though -- they don't use atomic operations with proper acquire/release memory ordering.

Reply:
The state-machine itself is not thread-safe. The `const` qualifiers on a couple of the methods are there only to indicate that they will not mutate internal state *other* than the transitionLockCount. I'll have a clarifying comment added